File: cpp/src/adtree.cc
#include "adtree.h"

using namespace std;
using namespace adtree;

ADNode::ADNode(dimension_t base, dimension_t num_dimensions, count_t count)
{
    this->base = base;
    this->count = count;
    this->num_children = num_dimensions - base;
    children = new VaryNode*[num_children];
    
    for (dimension_t i=0; i<num_children; i++)
    {
        children[i] = NULL;
    }
}

ADNode::~ADNode()
{
    for (dimension_t i=0; i<num_children; i++)
    {
        if (children[i] != NULL)
        {
            delete children[i];
        }
    }
    
    delete [] children;
}


VaryNode * ADNode::get_child(dimension_t i)
{
    return children[i-base];
}

void ADNode::set_child(dimension_t i, VaryNode *vn)
{
    children[i-base] = vn;
}


VaryNode::~VaryNode()
{
    for (auto it = children.begin(); it != children.end(); it++)
    {
        delete it->second;
    }
}


ADNode * VaryNode::get_child(string k)
{
    auto p = children.find(k);
    
    if (p == children.end())
    {
        return NULL;
    }
    
    return p->second;
}


void VaryNode::set_child(string k, ADNode *ad)
{
    children[k] = ad;
}


ADTree::ADTree(std::string * dimensions, dimension_t num_dimensions, recordlist_t records)
{
    this->dimensions = dimensions;
    this->num_dimensions = num_dimensions;
    root = make_adnode(0, records);
}


ADTree::~ADTree()
{
    delete root;
}


ADNode * ADTree::make_adnode(dimension_t i, recordlist_t records)
{
    auto ad = new ADNode(i, num_dimensions, records.size());
    
    for (dimension_t j=i; j<num_dimensions; j++)
    {
        ad->set_child(j, make_varynode(j, records));
    }
    
    return ad;
}


VaryNode * ADTree::make_varynode(dimension_t i, recordlist_t records)
{
    map<string,recordlist_t*> records_by_value;
    
    for (auto it = records.begin(); it != records.end(); it++)
    {
        recordlist_t *list;
        std::string * r = *it;
        string k = r[i];
        
        auto p = records_by_value.find(k);
        
        if (p != records_by_value.end())
        {
            list = p->second;
        }
        else
        {
            list = new recordlist_t;
            records_by_value[k] = list;
        }
        
        list->push_back(r);
    }
    
    string mcv;
    count_t count = 0;
    
    for (auto p = records_by_value.begin(); p != records_by_value.end(); p++)
    {
        if (p->second->size() > count)
        {
            count = p->second->size();
            mcv = p->first;
        }
    }
    
    auto vn = new VaryNode;
    vn->mcv = mcv;
    
    for (auto p = records_by_value.begin(); p != records_by_value.end(); p++)
    {
        if (p->first != mcv)
        {
            vn->set_child(p->first, make_adnode(i+1, *(p->second)));
        }
        
        delete p->second;
    }
    
    return vn;
}


count_t ADTree::count(query_t query)
{
    vector<Condition*> conditions;
    
    for (dimension_t i=0; i<num_dimensions; i++)
    {
        auto p = query.find(dimensions[i]);
        
        if (p != query.end())
        {
            auto condition = new Condition;
            condition->i = i;
            condition->value = p->second;
            conditions.push_back(condition);
        }
    }
    
    if (conditions.size() == 0)
    {
        return root->count;
    }
    
    count_t c = count(root, conditions, 0);
    
    for (auto it = conditions.begin(); it != conditions.end(); it++)
    {
        delete (*it);
    }
    
    return c;
}

count_t ADTree::count(ADNode *ad, std::vector<Condition*> conditions, dimension_t i)
{
    if (i == conditions.size())
    {
        return ad->count;
    }
    
    auto vn = ad->get_child(conditions[i]->i);
    
    if (conditions[i]->value.compare(vn->mcv) == 0)
    {
        count_t c = count(ad, conditions, i+1);
        
        for (auto p = vn->children.begin(); p != vn->children.end(); p++)
        {
            auto _c = count(p->second, conditions, i+1);
            c -= _c;
        }
        
        return c;
    }
    
    auto cad = vn->get_child(conditions[i]->value);
    
    if (cad == NULL)
    {
        return 0;
    }
    
    return count(cad, conditions, i+1);
}


rows_t ADTree::table(columns_t columns)
{
    vector<dimension_t> columns_i;
    
    for (dimension_t i=0; i<num_dimensions; i++)
    {
        for (auto it = columns.begin(); it != columns.end(); it++)
        {
            if (dimensions[i].compare(*it) == 0)
            {
                columns_i.push_back(i);
            }
        }
    }
    
    if (columns_i.size() == 0)
    {
        rows_t rows;
        return rows;
    }
    
    return table(root, columns_i, 0);
}

rows_t ADTree::table(ADNode *ad, vector<dimension_t> columns, dimension_t i)
{
    rows_t rows;
    
    if (i == columns.size())
    {
        auto row = new ResultRow(columns.size());
        row->count = ad->count;
        rows.push_back(row);
        return rows;
    }
    
    auto vn = ad->get_child(columns[i]);
    count_t total = 0;
    
    for (auto p = vn->children.begin(); p != vn->children.end(); p++)
    {
        auto next_rows = table(p->second, columns, i+1);
        
        for (auto it = next_rows.begin(); it != next_rows.end(); it++)
        {
            auto row = *it;
            total += row->count;
            row->columns[i] = p->first;
            rows.push_back(row);
        }
    }
    
    rows_t next_rows = table(ad, columns, i+1);
    
    if (columns.size()-1 != i)
    {
        for (auto it_a = next_rows.begin(); it_a != next_rows.end(); it_a++)
        {
            auto m = *it_a;
            
            for (auto it_b = rows.begin(); it_b != rows.end(); it_b++)
            {
                auto s = *it_b;
                bool match = true;
                
                for (dimension_t j=i+1; j < columns.size(); j++)
                {
                    if (m->columns[j].compare(s->columns[j]) != 0)
                    {
                        match = false;
                        break;
                    }
                }
                
                if (match)
                {
                    m->count -= s->count;
                }
            }
            
            if (m->count != 0)
            {
                m->columns[i] = vn->mcv;
                rows.push_back(m);
            }
        }
    }
    else
    {
        for (auto it = next_rows.begin(); it != next_rows.end(); it++)
        {
            auto row = *it;
            row->count -= total;
            
            if (row->count != 0)
            {
                row->columns[i] = vn->mcv;
                rows.push_back(row);
            }
        }
    }
    
    return rows;
}

void ADTree::delete_rows(rows_t rows)
{
    for (auto it = rows.begin(); it != rows.end(); it++)
    {
        delete [] (*it)->columns;
    }
}

File: python/adtree.py
import time

yapgvb = None
try:
    import yapgvb
except ImportError:
    pass

class ADNode(object):
    
    def __init__(self, count):
        self.count = count
        self.vary_nodes = {}
        
    def get(self, i):
        return self.vary_nodes.get(i)
    
    
class VaryNode(object):
    
    def __init__(self):
        self.mcv = None
        self.adnodes = {}
        
    def get(self, a):
        return self.adnodes.get(a)
    
    
class ADTree(object):
    
    def __init__(self, dimensions, records):
        self.dimensions = dimensions
        self.records = self._translate_records(records)
        self.root = self.make_adnode(0, self.records)
        self._node_id = 0
        
    def get_record(self, i):
        return self.records[i]
        
    def _translate_records(self, records):
        newrecords = []
        
        for r in records:
            newr = []
            for k in self.dimensions:
                val = r.get(k,'')
                if val is None:
                    val = ''
                newr.append(val)
            newrecords.append(newr)
        
        return newrecords
        
    def make_adnode(self, i, records):
        n = ADNode(len(records))
        
        for j in range(i, len(self.dimensions)):
            n.vary_nodes[j] = self.make_varynode(j, records)
        
        return n
        
        
    def make_varynode(self, i, records):
        v = VaryNode()
        child_records = {}
        
        for j in records:
            if j[i] not in child_records:
                child_records[j[i]] = [j]
            else:
                child_records[j[i]].append(j)
        
        v.mcv = max(child_records, key=lambda k: len(child_records[k]))
        
        for k,rec in child_records.items():
            if k == v.mcv:
                continue
            else:
                v.adnodes[k] = self.make_adnode(i+1, rec)
                
        return v
        
    def count(self, **kwargs):
        q = []
        
        for k in self.dimensions:
            q.append(kwargs.get(k,None))
        
        query = []
        
        for i,n in enumerate(q):
            if n is None:
                continue
            else:
                query.append((i,n))
        
        return self._count(self.root, query)
        
        
    def _count(self, adnode, query):
        if not query:
            return adnode.count
        
        vn = adnode.get(query[0][0])
        
        if query[0][1] == vn.mcv:
            c = self._count(adnode, query[1:])
            
            for k,ad in vn.adnodes.items():
                c -= self._count(ad, query[1:])
            
            return c
            
        ad = vn.get(query[0][1])
        
        if not ad:
            return 0
        return self._count(ad, query[1:])
        
        
    def table(self, *q):
        query = []
        
        for i,k in enumerate(self.dimensions):
            for a in q:
                if a == k:
                    query.append(i)
        
        return self._make_table(self.root, query)
        
        
    def _make_table(self, adnode, query):
        if not query:
            return [[adnode.count]]
        
        vn = adnode.get(query[0])
        rows = []
        total = 0
        
        for k,ad in vn.adnodes.items():
            for r in self._make_table(ad, query[1:]):
                rows.append([k] + r)
                total += r[-1]
        
        mcv_rows = []
        
        for m in self._make_table(adnode, query[1:]):
            if len(m) == 1:
                mcv_rows.append([vn.mcv, m[-1] - total])
            else:
                match = False
                for s in rows:
                    if m[:-1] == s[1:-1]:
                        m[-1] -= s[-1]
                mcv_rows.append([vn.mcv]+m)
        
        return rows + [r for r in mcv_rows if r[-1] > 0]
        
        
    def query(self, *q, **c):
        query = []
        conditions = []
        
        for i,k in enumerate(self.dimensions):
            v = c.get(k)
            if v is not None:
                conditions.append((i,v))
            else:
               for a in q:
                    if a == k:
                        query.append(i)
        
        return self._query(self.root, query, conditions)
        
    
    def _query(self, adnode, query, conditions):
        if not conditions:
            return self._make_table(adnode, query)
        
        if not query:
            return [[self._count(adnode, conditions)]]
        
        if query[0] < conditions[0][0]:
            vn = adnode.get(query[0])
            rows = []
            total = 0
            
            for k,ad in vn.adnodes.items():
                for r in self._query(ad, query[1:], conditions):
                    if r[-1] > 0:
                        rows.append([k] + r)
                        total += r[-1]
            
            mcv_rows = []
            
            for m in self._query(adnode, query[1:], conditions):
                if m[-1] > 0:
                    if len(m) == 1:
                        mcv_rows.append([vn.mcv, m[-1] - total])
                    else:
                        for s in rows:
                            if m[:-1] == s[1:-1]:
                                m[-1] -= s[-1]
                        mcv_rows.append([vn.mcv]+m)
            
            return rows + [r for r in mcv_rows if r[-1] > 0]
        else:
            vn = adnode.get(conditions[0][0])
            
            if conditions[0][1] == vn.mcv:
                rows = []
                total = 0
                
                for k,ad in vn.adnodes.items():
                    for r in self._query(ad, query, conditions[1:]):
                        if r[-1] > 0:
                            rows.append(r)
                            total += r[-1]
                            
                mcv_rows = []
                
                for m in self._query(adnode, query, conditions[1:]):
                    if m[-1] > 0:
                        if len(m) == 1:
                            mcv_rows.append([vn.mcv, m[-1] - total])
                        else:
                            for s in rows:
                                if m[:-1] == s[:-1]:
                                    m[-1] -= s[-1]
                            mcv_rows.append(m)
                
                return [r for r in mcv_rows if r[-1] > 0]
            
            ad = vn.get(conditions[0][1])
            
            if not ad:
                return []
            
            return self._query(ad, query, conditions[1:])
        
        
    def pprint(self):
        self._print('NULL', self.root, 0)
            
    def _print(self, val, adnode, depth):
        print (depth*"  ")+"%s (%s)" % (val, adnode.count)
        depth += 1
        for i,n in adnode.vary_nodes.items():
            print (depth*"\t")+"a%s (mcv=%s)" % (i, n.mcv)
            for k,an in n.adnodes.items():
                if an is not None:
                    self._print(k, an, depth+1)
    
    def draw(self, path='adtree.png'):
        G = yapgvb.Digraph('ADTree')
        label = self._node_id_iter()
        viz_adnode = G.add_node(label.next(),label='root (%s)' % self.root.count, shape=yapgvb.shapes.circle, color = yapgvb.colors.blue, root=True)
        self._build_graph(G, self.root, viz_adnode, label)
        G.layout(yapgvb.engines.dot)
        G.render(path)
        
        
    def _build_graph(self, G, adnode, viz_adnode, label):
        for i,vn in adnode.vary_nodes.items():
            viz_vn = G.add_node(label.next(), label='%s (mcv=%s)' % (self.dimensions[i],vn.mcv), shape=yapgvb.shapes.circle, color = yapgvb.colors.red)
            viz_adnode >> viz_vn
            for k,an in vn.adnodes.items():
                viz_an = G.add_node(label.next(),label="%s (%s)" % (k, an.count), shape=yapgvb.shapes.circle, color = yapgvb.colors.blue)
                viz_vn >> viz_an
                self._build_graph(G, an, viz_an, label)
    
    def _node_id_iter(self):
        i = 0
        while 1:
            yield str(i)
            i+=1
            
            
    def size(self):
        size = {
            'dimensions': len(self.dimensions),
            'records': len(self.records),
            'record_pointers': 0,
            'adnodes': 0,
            'varynodes': 0
        }
        
        self._size(self.root, size)
        return size
        
    def _size(self, adnode, size):
        size['adnodes'] += 1
        size['record_pointers'] += adnode.count
        
        for i,vn in adnode.vary_nodes.items():
            size['varynodes'] += 1
            for k,an in vn.adnodes.items():
                if an:
                    self._size(an, size)
            
            
if __name__ == "__main__":
    from pprint import pprint
    
    fields = ('A', 'B', 'C')
    
    records = [
        ('foo', 'bar', 'baz'),
        ('goo', 'gar', 'gaz'),
        ('hoo', 'har', 'haz'),
        ('moo', 'mar', 'maz'),
        ('foo', 'zar', 'baz'),
        ('goo', 'zar', 'gaz'),
        ('hoo', 'zar', 'haz'),
        ('hoo', 'zar', 'paz'),
        ('coo', 'zar', 'maz')
    ]
    
    records = [dict([(k,v) for k,v in zip(fields,r)]) for r in records]
    
    tree = ADTree(fields, records)
    print
    pprint(tree.count(A='foo'))
    print '---'
    pprint(tree.count(B='zar', C='gaz'))
    print '---'
    pprint(tree.table('A','B'))
    print '---'
    pprint(tree.table('A','C'))
    print '---'
    pprint(tree.query('A','B', C='haz'))